pkg/image/registry/imagestream/etcd/image.go

```diff
@@ -0,0 +1,224 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+
+	imagev1 "github.com/openshift/api/image/v1"
+	imageapi "github.com/openshift/origin/pkg/image/apis/image"
+)
+
+// ImageLayerIndex is a cache of image digests to the layers they contain.
+// Because a very large number of images can exist on a cluster, we only
+// hold in memory a small subset of the full image object.
+type ImageLayerIndex interface {
+	HasSynced() bool
+	GetByKey(key string) (item interface{}, exists bool, err error)
+	Run(stopCh <-chan struct{})
+}
+
+type ImageListWatch interface {
+	List(metav1.ListOptions) (*imagev1.ImageList, error)
+	Watch(metav1.ListOptions) (watch.Interface, error)
+}
+
+type imageLayerIndex struct {
+	informer cache.SharedIndexInformer
+}
+
+func NewEmptyLayerIndex() ImageLayerIndex {
+	return imageLayerIndex{}
+}
+
+func (i imageLayerIndex) HasSynced() bool {
+	if i.informer == nil {
+		return true
+	}
+	return i.informer.HasSynced()
+}
+func (i imageLayerIndex) GetByKey(key string) (item interface{}, exists bool, err error) {
+	if i.informer == nil {
+		return nil, false, nil
+	}
+	return i.informer.GetStore().GetByKey(key)
+}
+func (i imageLayerIndex) Run(stopCh <-chan struct{}) {
+	if i.informer == nil {
+		return
+	}
+	i.informer.Run(stopCh)
+}
+
+// NewImageLayerIndex creates a new index over a store that must return
+// images.
+func NewImageLayerIndex(lw ImageListWatch) ImageLayerIndex {
+	informer := cache.NewSharedIndexInformer(&cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			list, err := lw.List(metav1.ListOptions{
+				ResourceVersion: options.ResourceVersion,
+				Limit:           options.Limit,
+				Continue:        options.Continue,
+			})
+			if err != nil {
+				return nil, err
+			}
+			// reduce the full image list to a smaller subset.
+			out := &metainternalversion.List{
+				Items: make([]runtime.Object, len(list.Items)),
+			}
+			out.Continue = list.Continue
+			out.ResourceVersion = list.ResourceVersion
+			for i, image := range list.Items {
+				out.Items[i] = &ImageLayers{
+					Name:     image.Name,
+					Layers:   image.DockerImageLayers,
+					Manifest: manifestFromImage(&image),
+				}
+			}
+			return out, nil
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			w, err := lw.Watch(metav1.ListOptions{
+				ResourceVersion: options.ResourceVersion,
+			})
+			if err != nil {
+				return nil, err
+			}
+			return watch.Filter(w, func(in watch.Event) (out watch.Event, keep bool) {
+				if in.Object == nil {
+					return in, true
+				}
+				// reduce each object to the minimal subset we need for the cache
+				image, ok := in.Object.(*imagev1.Image)
+				if !ok {
+					return in, true
+				}
+				layers := &ImageLayers{
```

Comment:
one function that takes an image and produces an imagelayers.  These implementations have already drifted.  When you get it in the list, you don't get a resource version.